Shark Cage/createACLinNewDesktop/createACLinNewDesktop/util.cpp

```diff
@@ -217,4 +248,89 @@ void enumerateSidsAndHashes(PTOKEN_ACCESS_INFORMATION pToken) {
 		LocalFree(stringSid);
 		getchar();
 	}
-}
\ No newline at end of file
+}
+
+//adopted from MSDN example
+BOOL setPrivilege(
+	HANDLE hToken,          // access token handle
+	LPCTSTR lpszPrivilege,  // name of privilege to enable/disable
+	BOOL bEnablePrivilege   // to enable or disable privilege
+)
+{
+	TOKEN_PRIVILEGES tp;
+	LUID luid;
+
+	if (!LookupPrivilegeValue(
+		NULL,            // lookup privilege on local system
+		lpszPrivilege,   // privilege to lookup
+		&luid))        // receives LUID of privilege
+	{
+		printf("LookupPrivilegeValue error: %u\n", GetLastError());
+		return FALSE;
+	}
+
+	tp.PrivilegeCount = 1;
+	tp.Privileges[0].Luid = luid;
+	if (bEnablePrivilege)
+		tp.Privileges[0].Attributes = SE_PRIVILEGE_ENABLED;
+	else
+		tp.Privileges[0].Attributes = 0;
+
+	// Enable the privilege or disable all privileges.
+
+	if (!AdjustTokenPrivileges(
+		hToken,
+		FALSE,
+		&tp,
+		sizeof(TOKEN_PRIVILEGES),
+		(PTOKEN_PRIVILEGES)NULL,
+		(PDWORD)NULL))
+	{
+		printf("AdjustTokenPrivileges error: %u\n", GetLastError());
+		return FALSE;
+	}
+
+	if (GetLastError() == ERROR_NOT_ALL_ASSIGNED)
+
+	{
+		printf("The token does not have the specified privilege. \n");
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+bool changeTokenCreationPrivilege(bool privilegeStatus) {
+	DWORD bufferSize = 0;
+	GetUserName(NULL, &bufferSize);
+	LPTSTR pUser_name = (LPTSTR) new BYTE[bufferSize];
+	GetUserName(pUser_name, &bufferSize);
+	wprintf(L"User account accessed: %s\n", pUser_name);
+
+	HANDLE current_process_handle;
+	HANDLE user_token_h;
+	current_process_handle = GetCurrentProcess();
+	if (!OpenProcessToken(current_process_handle, TOKEN_ALL_ACCESS, &user_token_h)) {
+		wprintf(L"Error getting token for privilege escalation\n");
+		return false;
+	}
+	return setPrivilege(user_token_h, SE_CREATE_TOKEN_NAME, privilegeStatus);
+}
+
+bool addGroupToTokenGroups(PSID sid, tokenStructures &tokenDeconstructed) {
+	PTOKEN_GROUPS tokenGroups = tokenDeconstructed.TokenGroups;
+	DWORD groupCount = tokenGroups->GroupCount;
+	SID_AND_ATTRIBUTES newGroup{ sid, SE_GROUP_ENABLED };
+
+	PTOKEN_GROUPS tokenGroupsMod = (PTOKEN_GROUPS)malloc(FIELD_OFFSET(TOKEN_GROUPS, Groups[groupCount+1]));
+	for (size_t i = 0; i < groupCount; i++)
+	{
+		tokenGroupsMod->Groups[i] = tokenGroups->Groups[i];
+	}
+	tokenGroupsMod->Groups[groupCount] = newGroup;
+	tokenGroupsMod->GroupCount = groupCount + 1;
+
+	tokenDeconstructed.TokenGroups = tokenGroupsMod;

```

Comment:
the memory of the old `TokenGroups` struct is leaked here

Reply:
Yeah about that... this was a bit tricky, since the old memory cannot be cleared up at that point since NtCreateToken is doing some magic and somehow referencing this memory... I redid it a different way so now it sould be cool 😎